client/my-sites/stats/feedback/index.tsx

```diff
@@ -175,27 +172,33 @@ function StatsFeedbackController( { siteId }: FeedbackProps ) {
 		}
 	}, [ isPending, isError, shouldShowFeedbackPanel ] );
 
-	const dismissPanelWithDelay = () => {
-		// Allows the animation to run first.
-		setTimeout( () => {
-			setIsFloatingPanelOpen( false );
-		}, FEEDBACK_PANEL_ANIMATION_DELAY_EXIT );
-	};
+	function dismissFloatingPanel() {
+		const delay = isFloatingPanelOpen ? FEEDBACK_PANEL_ANIMATION_DELAY_EXIT : 0;
+		setAnimationName( FEEDBACK_PANEL_ANIMATION_NAME_EXIT );
+		return new Promise( ( resolve ) => {
+			setTimeout( () => {
+				setIsFloatingPanelOpen( false );
+				resolve( '' );
+			}, delay );
+		} );
+	}
```

Comment:
The code here is very confusing. None of the function called here is `aync` but it is wrapped with Promise. That said, it is not guaranteed that the the panel is closed on return.

Reply:
This function was put together following the example from MDN on [using promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises). The goal was to have the `dismiss.then()` pattern (like the fetch API) so it was clear that follow up actions happened once the panel was closed.

My understanding is we do not need an `async` label here as we aren't using the async/await paradigm. Unless I'm misunderstanding things. 

> That said, it is not guaranteed that the the panel is closed on resolve.

How do you mean? I think it is guaranteed to be closed once we call `setIsFloatingPanelOpen( false )`.

Reply:
JavaScript is a single threaded language, and it's not possible to create async actions ourselves, which means the function is not provided by browser or system, we are not able to create real async functions, because they resolve instantly.

In the code above, what's actually working is just the timeout, and that is all. It doesn't wait for the panel to close, because the operation here is just not async. Even if it is, we should either use `await` or chain it with `.then` to be able to carry on the actions after promise resolve.

With the following code, it could achieve the same result:

```JavaScript
	useEffect( () => {
		// if ( ! isPending && ! isError && shouldShowFeedbackPanel ) {
		setTimeout( () => {
			setIsFloatingPanelOpen( true );
		}, FEEDBACK_PANEL_PRESENTATION_DELAY );
		// }
	}, [ isPending, isError, shouldShowFeedbackPanel ] );

	function dismissFloatingPanel() {
		const delay = isFloatingPanelOpen ? FEEDBACK_PANEL_ANIMATION_DELAY_EXIT : 0;
		setAnimationName( FEEDBACK_PANEL_ANIMATION_NAME_EXIT );
		return setTimeout( () => {
			setIsFloatingPanelOpen( false );
		}, delay );
	}

	const handleButtonClick = ( action: string ) => {
		switch ( action ) {
			case ACTION_SEND_FEEDBACK:
				dismissFloatingPanel();
				setTimeout( () => setIsOpen( true ), FEEDBACK_PANEL_ANIMATION_DELAY_EXIT );

				break;
			case ACTION_DISMISS_FLOATING_PANEL:
				dismissFloatingPanel();
				updateFeedbackPanelHibernationDelay();
				trackStatsAnalyticsEvent( `stats_feedback_${ ACTION_DISMISS_FLOATING_PANEL }` );
				break;
			case ACTION_LEAVE_REVIEW:
				dismissFloatingPanel();
				window.open( FEEDBACK_LEAVE_REVIEW_URL );

				break;
			// Ignore other cases.
		}
	};
```

Reply:
I found this article quite helpful: https://www.freecodecamp.org/news/javascript-asynchronous-operations-in-the-browser/

Reply:
> JavaScript is a single threaded language, and it's not possible to create async actions ourselves, which means the function is not provided by browser or system, we are not able to create real async functions, because they resolve instantly.

This doesn't quite fit with my reading of the docs. My understanding is that Promises are the recommended way to wrap callback-based APIs like `setTimeout()` so they can be used in chaining. The MDN docs show two examples of this [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise), one which is specifically showing it as a wrapper for `setTimeout()`.

In terms of single threaded-ness, I believe I follow your point. I understand we aren't getting async execution here. Instead we are scheduling a callback with the event loop. The goal with wrapping the callback in a Promise here is to allow for use of the chaining syntax, which keeps the animation timing central to the single `dismissFloatingPanel()` function while allowing us run code after the animation.

The code you shared would accomplish this as well but I believe it to be more complicated. For example:

```
  case ACTION_SEND_FEEDBACK:
    dismissFloatingPanel();
    const delay = isFloatingPanelOpen ? FEEDBACK_PANEL_ANIMATION_DELAY_EXIT : 0;
    setTimeout( () => setIsOpen( true ), delay );
    break;
```

Vs:

```
  case ACTION_SEND_FEEDBACK:
    dismissFloatingPanel().then( () => {
      setIsOpen( true );
    } );
  break;
```

Note that we'd need the same timing-related code when handling the "leave review" action as well.

Reply:
> In the code above, what's actually working is just the timeout, and that is all.

I understand the point you are making here. The timeout is what makes this work. I just want to be clear that this is the recommended approach for wrapping callback-based APIs per [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise). If we don't want to use this approach, I'm fine with that. I just want to make sure I'm not misunderstanding something important here.

> It doesn't wait for the panel to close, because the operation here is just not async (I mean it is async but not really the case for setIsFloatingPanelOpen tho). Even if it is, we should either use await or chain it with .then to be able to carry on the actions after promise resolve.

I think the important notes here are:

1. If you capture and log the promise, it starts as pending, so the proper order of operations is observed.
2. The call to `setIsFloatingPanelOpen()` executes after the specified delay, meaning the animation should have the necessary time to complete.
3. We are making use of the `.then()` syntax at the call site to handle opening a new tab/window or present the feedback modal.

Also, just to be clear, I'm not trying to be difficult here. I'm happy to do it either way. I just want to make sure I'm understanding any nuances that I might be missing so I properly grok things going forward.

Reply:
> My understanding is that Promises are the recommended way to wrap callback-based APIs like setTimeout() so they can be used in chaining.

Yes you are right. I think I got what you were trying to achieve here now. I still don't quite like what's happening in here tho. If we are using the timeouts, then use timeouts, it's an estimation of how long the animation would take and then take next actions. The promise only saves setting up another timeout, but made the code harder to parse. Also `setIsFloatingPanelOpen` and `setIsOpen` could be in the same timer really.

```JavaScript
case ACTION_SEND_FEEDBACK:
 dismissFloatingPanel();
 const delay = isFloatingPanelOpen ? FEEDBACK_PANEL_ANIMATION_DELAY_EXIT : 0;
 setTimeout( () => { setIsOpen( true ); setIsFloatingPanelOpen( false ); }, delay );
 break;
```

My suggestion would be that, we either do it precisely - listen on animation start / end events or we do it simply - relying on timers, and getting rid of the promises.

Sorry to be a pain here. The reason I'm having long discussions here is that you are passionate about code quality and so am I, and this is a great opportunity to have a very specific and in-depth conversation. Plus the PR doesn't seem to be urgent. 

@dognose24 I wonder what you think?

Reply:
> Note that we'd need the same timing-related code when handling the "leave review" action as well.

It looks a little unintuitive to wait for the animation to open a new tab. 🤔

Reply:
> As the function dismissFloatingPanel is shared to achieve the delay, is passing a callback enough?

As Kevin mentioned, "this is the recommended approach for wrapping callback-based APIs per the MDN docs" I think 🤔